Agnostic Renderer

Zinston opened the pull request.

Goaman commented on packages/plugin-char/Char.ts:
```diff
@@ -42,6 +53,52 @@ export class Char extends JWPlugin {
+        context.parentNode.appendChild(fragment);
+        context.renderingMap = createMapFromDom([
+            [renderedNode, charNodes.concat(renderedFormats)],
+        ]);
```
Should rather return the renderingMap rather than writing the context

Goaman commented on packages/plugin-dom-renderer/DomRenderer.ts:
```diff
@@ -25,43 +27,77 @@ export class DomRenderer extends Renderer<DocumentFragment> {
+        });
+
+        // We're done using the rendering map: wipe it.
+        context.renderingMap = new Map<Node, VNode[]>();
```
If the renderingMap is returned, no need for this line

Goaman commented on packages/plugin-dom-renderer/DomRenderer.ts:
```diff
@@ -25,43 +27,77 @@ export class DomRenderer extends Renderer<DocumentFragment> {
+            renderedVNodes.forEach((node: VNode, index: number) => {
+                VDocumentMap.set(node, domNode, index);
+            });
+        });
```
Should have it's own map. Maybe in another PR?

dmo-odoo replied:
Yes, this is out of scope of this PR.

dmo-odoo commented on packages/core/src/JWEditor.ts:
```diff
@@ -140,7 +156,7 @@ export class JWEditor {
     execCommand(id: CommandIdentifier, args?: CommandArgs): void {
         this.dispatcher.dispatch(id, args);
-        this.renderer.render(this.vDocument, this.editable);
+        this.renderers[this.defaultRendererId].render(this.vDocument, this.editable);
```
All renderers should be called ? So if I have installed, say, a PDFRenderer, it re-generates a new PDF at every change ? I would have expected it to not do anything unless I click on "Export to PDF" or whatever. It looks like we may need a separation between "live" renderers and "exports" renderer ?

Goaman replied:
They should at least have a signal that tells them that the document has changed.
Maybe instead of calling a method `render`, we could call something like `changeDocument`.

Gorash replied:
We have a confusion of term / name.
I think that it is necessary to make the difference between the signal to render them (informing that they can update), and the request to the editor to obtain the result of a specific rendering.
The `renderers` (plugin thereafter) know if they are displayed or not, and know if they update.
Clicking on 'save as' will be another command.

dmo-odoo commented on packages/core/src/JWEditor.ts:
```diff
@@ -120,10 +121,18 @@ export class JWEditor {
+        // Register the parsing predicate of this plugin if it has any.
+        Object.keys(pluginClass.renderingPredicates).forEach(rendererId => {
+            if (Object.keys(this.renderers).includes(rendererId)) {
+                this.renderers[rendererId].addRenderingPredicate(
```
I guess we'll veer away from predicates here as well.

dmo-odoo commented on packages/core/src/Renderer.ts:
```diff
@@ -1,6 +1,25 @@
+export class Renderer<T> {
+    renderingPredicates = new Set<RenderingPredicate<T>>();
+    readonly name: string;
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function, @typescript-eslint/no-explicit-any
```
// dmo-disable-next-line @dmo-eslint/no-eslint-disable

Zinston replied:
Well. Agreed. But do you know how I can do without?

dmo-odoo commented on packages/plugin-dom-renderer/DomRendererPlugin.ts:
```diff
@@ -2,5 +2,5 @@ import { JWPlugin } from '../core/src/JWPlugin';
 
 export class DomRendererPlugin extends JWPlugin {
-    static readonly renderers = [DomRenderer];
+    static readonly renderers = [new DomRenderer()];
```
@Zinston Just so you know, if you make Renderer an interface (as suggested by nby) rather than a class, it would be possible to write:
```export class DomRendererPlugin extends JWPlugin implements Renderer {```

Such that a plugin can be a renderer, rather than serving no other purpose than providing a `renderers` key to another object.

You can even implements multiple interfaces, for example:
```export class MyPlugin extends JWPlugin implements Renderer, WhateverFeature, WhateverOtherFeature {```

Zinston replied:
This particular plugin only does that but nothing should stop you from making a plugin that declares renderers _and_ does other things too... Also, I made `Renderer` a class so that it could handle `addRenderingFunction` and we wouldn't have to rewrite it for each renderer.

dmo-odoo commented on packages/core/src/VDocumentMap.ts:
```diff
@@ -92,3 +92,21 @@ export function createMap(items: Array<[VNode, Node | Node[]]>): Map<VNode, Node
+        }
+    });
+    return map;
+}
```
I think it's just an artifact of the previous `createMap` paradigm of the parser branch on which this branch was based. We discussed about it yesterday and decided to remove it from the parser branch so of course it should be removed from here as well.

dmo-odoo commented on packages/plugin-dom-renderer/DomRenderer.ts:
```diff
@@ -1,14 +1,16 @@
 import { VElement } from '../core/src/VNodes/VElement';
 
-export class DomRenderer extends Renderer<DocumentFragment> {
+export interface DomRenderingContext extends AbstractRenderingContext {
```
Yeah, actually, I would have moved nearly everything from Renderer.ts to DomRenderer.ts and just keep the "render" method in the Renderer interface. If something else can be generalized later, we can do it at that point, but now it's a bit too early and erroneous.

Zinston replied:
How do we define RenderingFunction then? Since each renderer can define it differently?

Zinston commented on packages/plugin-dom/DomRenderer.ts:
```diff
@@ -75,7 +75,10 @@ export class DomRenderer extends Renderer<
             const renderingResult = renderingFunction({ ...context });
             if (renderingResult) {
-                return renderingResult;
+                context = renderingResult[0];
```
This is why I call it "dirty make-it-work" ;-)
How should we handle this kind of situation? We need to be able to render a wrapper, then let the node render itself within said wrapper.
Also there is an issue of priorities. By chance right now it works, but it's only by chance: List has to take control whenever a ListNode's children have to be rendered. But if another renderingFunction renders the node, List will never be called. List _could_ handle the rendering of all its children, like it was done in the non-agnostic renderer... but then they need to be able to call `render` on each of said children.

Zinston commented:
@dmo-odoo Ok so I'm finally done rebasing and updating this for lists. I also updated to follow the new patterns of Parser (like the stack, and no predicates), and a few of the review comments. Now I added a few comments myself because some things require discussion. There is still work on this PR. It's no easy job but it's prioritary right now. Maybe we can finish this in tandem?

dmo-odoo commented on packages/core/src/JWEditor.ts:
```diff
@@ -21,10 +23,11 @@ export class JWEditor {
     eventManager: EventManager;
     pluginsRegistry: JWPlugin[];
-    renderer: Renderer;
+    renderers: Record<string, Renderer> = {};
```
I suggest to give a nickname to that `string`. Maybe `RenderingOutputType` ? That's a bit long, though.

dmo-odoo commented on packages/core/src/JWEditor.ts:
```diff
@@ -111,6 +116,14 @@ export class JWEditor {
+        // will replace the previous one.
+        if (pluginClass.renderers) {
+            pluginClass.renderers.forEach(Renderer => {
+                this.renderers[Renderer.id] = new Renderer();
```
sproutch ! 🔨

Zinston commented on packages/core/src/JWEditor.ts:
```diff
@@ -11,6 +12,7 @@ import { DevTools } from '../../plugin-devtools/src/DevTools';
 export interface JWEditorConfig {
     debug?: boolean;
     theme?: string;
+    defaultRendererId?: string;
```
We don't. It sucks. But I haven't managed without yet.

dmo-odoo commented on packages/plugin-list/List.ts:
```diff
@@ -108,60 +108,69 @@ export class List extends JWPlugin {
             context = result[0];
-        }
-        if (node.is(ListNode)) {
+            this._isRenderingLi = false;
```
By "remove list rendering hack", you mean "replace it by another hack", right ? 🚎

Zinston replied:
![image](https://user-images.githubusercontent.com/24205914/73541293-58609680-4432-11ea-81f9-b85afc8ccf15.png)

dmo-odoo commented on packages/core/src/Renderer.ts:
```diff
@@ -1,11 +1,12 @@
 
 export type RendererIdentifier = string;
-export type RenderingFunction<Context = {}, T = {}> = (context: Context) => T;
+export type RenderingFunction<To = {}> = (node: VNode, to: To) => Map<VNode, To[]>;
```
That is kinda weird... If you want to render a node into string you have to pass it a string ?

dmo-odoo commented on packages/plugin-dom/DomRenderer.ts:
```diff
@@ -40,68 +29,61 @@ export class DomRenderer extends Renderer<
-        const firstVNode = renderedRoot ? root.firstChild() : root;
-
-        if (!firstVNode) return;
+        const domRoot = VDocumentMap.toDom(root);
```
Is it the root of the dom ?

dmo-odoo commented on packages/plugin-dom/DomRenderer.ts:
```diff
@@ -40,68 +29,61 @@ export class DomRenderer extends Renderer<
+        let rendered = [];
+        while (node) {
+            if (!rendered.includes(node)) {
+                rendered = [...rendered, ...this.renderNode(node, domParent).keys()];
```
I would explicitely call `renderNode` in its own statement, it is in fact the most important part of the algorithm.

dmo-odoo commented on packages/plugin-dom/DomRenderer.ts:
```diff
@@ -40,68 +29,61 @@ export class DomRenderer extends Renderer<
+                rendered = [...rendered, ...this.renderNode(node, domParent).keys()];
+            }
+            node = node.nextSibling();
+        }
```
Rather than a let, a while and another reasignation, I think you could just have done:
```
for (const child of root.children) {
    ...
}
```

dmo-odoo commented on packages/plugin-dom/DomRenderer.ts:
```diff
@@ -122,54 +104,16 @@ export class DomRenderer extends Renderer<
-        return this._contextStack[this._contextStack.length - 1];
+        });
+
+        domParent.appendChild(fragment);
```
I think the very point of backing down to a "rendering a node renders its children" mechanism is to actually do the append in the parent itself. So the api of `render` would just be `render<T>(node: VNode): T` and then the parent would take the result and append it where he wants it to be appended. The point is the parent controls where the children is placed.

dmo-odoo commented on packages/plugin-dom/DomRenderer.ts:
```diff
@@ -122,54 +104,16 @@ export class DomRenderer extends Renderer<
+        let rendered = [];
+        node.children.forEach(child => {
+            if (!rendered.includes(child)) {
+                rendered = [...rendered, ...this.renderNode(child, renderedElement).keys()];
```
Yeah, really that's not ok :p Actually, with this mechanism, is the map still needed ? I think it is yes, but we said we would set it manually. That's something else though.

dmo-odoo commented on packages/plugin-list/List.ts:
```diff
@@ -102,74 +101,66 @@ export class List extends JWPlugin {
+                // Render the node's children
+                let rendered = [];
+                node.children.forEach(child => {
+                    if (!rendered.includes(child)) {
```
How could this happen ?

dmo-odoo commented on packages/plugin-list/List.ts:
```diff
@@ -102,74 +101,66 @@ export class List extends JWPlugin {
-                currentContext = this.editor.renderers.dom.renderNode({
-                    ...currentContext,
-                }) as DomRenderingContext;
+                this.editor.renderers.dom.renderNode(node, domListItem) as Map<VNode, Node[]>;
```
as

dmo-odoo commented on packages/plugin-list/List.ts:
```diff
@@ -102,74 +101,66 @@ export class List extends JWPlugin {
+                const br = document.createElement('BR');
+                domListItem.appendChild(br);
+                domNodes.push(br);
+            }
```
In my understanding, the one and only time where you would still have to do that is in the case where you render a `P` as a `LI` because it is a child of a list and that particular `P` is empty. It's the only time where you "prevent" somewhat the node from rendering itself using the standard mechanism.

Zinston replied:
So you want me to move this into the above `if`?

dmo-odoo commented on packages/core/src/EventManager.ts:
```diff
@@ -25,7 +25,7 @@ export class EventManager {
      * Callback given to the normalizer.
      */
-    _onNormalizedEvent(customEvent: CustomEvent): void {
+    async _onNormalizedEvent(customEvent: CustomEvent): Promise<void> {
```
You put `async` and `Promise` but no `await`, how does that work ? I see most of the time we `return this.editor.execCommand` so I guess it works, but the `default` case does not `return`.

Zinston replied:
We return an asynchronous function call so that checks out. The default case I'm not sure...

dmo-odoo commented on packages/plugin-char/test/Char.test.ts:
```diff
@@ -8,17 +8,20 @@ import { Parser } from '../../core/src/Parser';
+const applyFormat = async (
+    editor: JWEditor,
+    format: 'bold' | 'italic' | 'underline',
+): Promise<void> => {
```
So much git conflits ahead... 👨‍💻 😵

Zinston replied:
:sos:

dmo-odoo commented on packages/plugin-dom/DomRenderer.ts:
```diff
@@ -78,12 +79,21 @@ export class DomRenderer extends Renderer<Node> {
+        let rendered = [];
+        for (const child of node.children) {
+            if (!rendered.includes(child)) {
+                const mapping = await this.renderNode(child, domParent);
```
Thank you.

dmo-odoo commented on packages/plugin-list/List.ts:
```diff
@@ -111,57 +111,47 @@ export class List extends JWPlugin {
-                mapping = [...mapping, ...this.renderListItemToDom(child, domListNode)];
-            });
+            for (const child of node.children) {
+                const childrenMapping = await this.renderListItemToDom(child, domListNode);
```
You mean `childMapping`, right ?

dmo-odoo commented on packages/plugin-list/List.ts:
```diff
@@ -23,80 +24,135 @@ function _isEmptyTextNode(node: Node): boolean {
+            await this.editor.renderers.dom.renderChildren(node, domListItem);
+        } else {
+            // The node was wrapped in a "LI" but needs to be rendered as well.
+            await this.editor.renderers.dom.renderNode(node, domListItem);
```
The `renderNode` function is on the renderer, right ? What if it was this function which does the mapping ? It would map the result of the rendering function to the `VNode` it got in argument. No more need for manual mappings *unless* you create unrepresented nodes. Example:
```typescript
// In DomRenderer.ts
renderNode(node: VNode) {
    const result = renderingFunction(node, this.domMap);
    if (result) {
        this.domMap.set(node, result);
    }
}

// In List.ts
renderListToDom(node: ListNode, domMap: Map<VNode: Node[]>) {
    const ul = createUL();
    for (const child of node) {
        const li = createLI();
        const domChild = this.editor.render(child);
        li.appendChild(domChild);
        domMap.get(child).unshift(li);
        ul.appendChild(li);
    }
}
```

Zinston replied:
Thanks for using ```typescript.

Zinston commented on packages/core/src/JWEditor.ts:
```diff
@@ -114,23 +114,29 @@ export class JWEditor {
+                if (Array.isArray(renderingFunctions)) {
+                    renderingFunctions.forEach(renderingFunction => {
+                        renderer.renderingFunctions.add(renderingFunction);
+                    });
```
No.
